### PR TITLE
Move GCM webhooks to dedicated Admission Controller

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area scalability
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR moves Gardener-Controller-Manager related webhooks to a dedicated component/binary called `Gardener-Admission-Controller`. Introducing a dedicated webhook server allows us to efficiently run multiple replicas in the garden cluster and also follows a separation of concerns.

Prerequisite for #2781

**Special notes for your reviewer**:
- I didn't remove the HTTPS settings for GCM because in the near future remaining endpoints should be protected by TLS (#2829)
- The namespace deletion validator now works with the cached runtime-client instead of client + project informer + shoot  informer.

**Release note**:

```noteworthy operator
Gardener validating webhooks have been moved from the `Gardener-Controller-Manager` to a dedicated component `Gardener-Admission-Controller`. Therefore, new values have been added to the Gardener Helm chart (`charts/gardener/controlplane`). Please consult the documentation (`docs/concepts/admission-controller.md`) for more information about the `Gardener-Admission-Controller`.
```
```noteworthy developer
Gardener validating webhooks have been moved from the `Gardener-Controller-Manager` to a dedicated component `Gardener-Admission-Controller`. Please run `make dev-setup` to create the necessary configuration for this component for local development use-cases. `make start-admission-controller` starts the new component locally.
```

### CODE REVIEW

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml
Comment (timebertt) [inline charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml]:
```suggestion
      {{- if gt (int .Values.global.admission.replicaCount) 1 }}
```

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/secret-cert.yaml
Comment (timebertt) [inline charts/gardener/controlplane/charts/runtime/templates/admission-controller/secret-cert.yaml:13]: Would it make sense, to use `kubernetes.io/tls` as the secret type here?
I see, we also don't use it in the other charts...
```suggestion
type: kubernetes.io/tls
```

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/serviceaccount.yaml
Comment (timebertt) [inline charts/gardener/controlplane/charts/runtime/templates/admission-controller/serviceaccount.yaml:2]: Does the admission-controller need any RBAC rules?

File: pkg/admissioncontroller/apis/config/types.go
Comment (timebertt) [inline pkg/admissioncontroller/apis/config/types.go]: Hmm, `unsecured, unauthenticated access` is not true here / might be confusing, right?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go:15]: License header missing

File: cmd/gardener-admission-controller/main.go
Comment (timebertt) [inline cmd/gardener-admission-controller/main.go:15]: License header missing

File: cmd/gardener-admission-controller/main.go
Comment (timebertt) [inline cmd/gardener-admission-controller/main.go]: Can you pass down the context via `ExecuteContext`? (You can then get it in `RunE` via `cmd.Context()`)
```suggestion
	command := app.NewCommandStartGardenerAdmissionController()

	if err := command.ExecuteContext(ctx); err != nil {
```

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go]: Can you return the errors instead of panicking?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go]: I think, this field can be unexported.
```suggestion
	configFile string
```

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go]:
```suggestion
		return fmt.Errorf("missing Gardener admission controller config file")
```

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go]: I am a bit confused by the version handling here...
Why are you first decoding the config in the internal version, then converting it back to the external one, default it and then convert it to the internal version again? Is this somehow required?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go:165]: Shouldn't the env var only be used as a fallback, if the config doesn't provide a kubeconfig, instead of overwriting the config?

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go]: Do we still need `admissionregistrationv1beta1`?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go:194]: I think, one problem that we might hit here, is that we now don't wait for the needed informers to sync before we start the webhook server.
One option would be to retrieve an informer for every resource we want to talk to and then use the `HasSynced` funcs to wait for the caches to sync.

WDYT about constructing the webhook handlers before starting the server, then passing the client to some kind of `Initialize` func there and let the handlers retrieve the needed informers and necessary `HasSynced` funcs?

Something like this would help to make sure, the webhook can directly serve requests off the cache, when it starts accepting traffic, and not to introduce to much unneeded latency, when calling the gardener api.

**Review by timebertt — CHANGES_REQUESTED** [review]
Nice PR, @timuthy! Very clean refactoring. Thanks!
I only have minor comments and nit suggestions :)

File: Dockerfile
Comment (rfranzke) [inline Dockerfile:58]: Just a reminder that we will need to adapt the `meta/ci` ref

File: docs/README.md
Comment (rfranzke) [inline docs/README.md]: Can you move it after the scheduler but before the seed admission controller?

File: docs/concepts/admission-controller.md
Comment (rfranzke) [inline docs/concepts/admission-controller.md]:
```suggestion
While the Gardener API server works with [admission plugins](./apiserver_admission_plugins.md) to validate and mutate resources belonging to Gardener related API groups, e.g. `core.gardener.cloud`, the same is needed for resources belonging to non-Gardener API groups as well, e.g. `Secret`s in the `core` API group.
```

File: example/20-componentconfig-gardener-admission-controller.yaml
Comment (rfranzke) [inline example/20-componentconfig-gardener-admission-controller.yaml]:
```suggestion
    port: 2721
```

File: hack/update-codegen.sh
Comment (rfranzke) [inline hack/update-codegen.sh]:
```suggestion
  --extra-peer-dirs=github.com/gardener/gardener/pkg/admissioncontroller/apis/config,github.com/gardener/gardener/pkg/admissioncontroller/apis/config/v1alpha1,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/conversion,k8s.io/apimachinery/pkg/runtime,k8s.io/component-base/config,k8s.io/component-base/config/v1alpha1 \
```

File: pkg/admissioncontroller/apis/config/v1alpha1/dafaults_test.go
Comment (rfranzke) [inline pkg/admissioncontroller/apis/config/v1alpha1/dafaults_test.go]: fix typo in file name

File: pkg/admissioncontroller/apis/config/v1alpha1/dafaults_test.go
Comment (rfranzke) [inline pkg/admissioncontroller/apis/config/v1alpha1/dafaults_test.go]:
```suggestion
			Expect(obj.Server.HTTPS.Port).To(Equal(2721))
```

File: pkg/admissioncontroller/apis/config/v1alpha1/defaults.go
Comment (rfranzke) [inline pkg/admissioncontroller/apis/config/v1alpha1/defaults.go]:
```suggestion
		obj.Server.HTTPS.Port = 2721
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/secret-cert.yaml
Comment (timuthy) [reply charts/gardener/controlplane/charts/runtime/templates/admission-controller/secret-cert.yaml:13]: Tbh, neither am I sure about the benefits or consequences nor have I found something during a quick research except that is has some benefits for `kubectl`. To be sure, I'd like to stick to `generic` for now since we use it in other charts as well if you agree?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timuthy) [reply cmd/gardener-admission-controller/app/gardener_admission_controller.go]: This is the way you apply the defaults registered in the scheme for this kind. We did the same before in GCM, so I didn't question this handling.

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timuthy) [reply cmd/gardener-admission-controller/app/gardener_admission_controller.go:165]: We do this for all of our components this way. Any objection to change it generally @rfranzke ?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timuthy) [reply cmd/gardener-admission-controller/app/gardener_admission_controller.go:194]: I was a bit more pragmatic and hope you are fine with the changes.

File: Dockerfile
Comment (timuthy) [reply Dockerfile:58]: Thanks, will do as soon as this PR is merged.

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/secret-cert.yaml
Comment (timebertt) [reply charts/gardener/controlplane/charts/runtime/templates/admission-controller/secret-cert.yaml:13]: Yep, agreed.

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (rfranzke) [reply cmd/gardener-admission-controller/app/gardener_admission_controller.go:165]: Let's simply keep it as it is/also was in the past to not break existing usages. It's not an important decision here, either you go left or right, and we decided to go left. :)

**Review by rfranzke — COMMENTED** [review]
No more comments from my side, I'll try to play with it locally later today.

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [reply cmd/gardener-admission-controller/app/gardener_admission_controller.go:194]: Hmm, ok. I would have to test that, but I would definitely expect some latency on startup in larger landscapes because the handler has to load all Shoots and all Projects. So, I am still a bit concerned about this.

Another option to mitigate this situation and still keep it pragmatic in code, would be to add a `startupProbe` (or `livenessProbe` with some `initialDelaySeconds`).
Still, this would only be based on heuristics and not really detect any long running initial syncs...
/cc @rfranzke

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml
Comment (timebertt) [inline charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml:83]: Can you add some health checks? I think we should at least add a TCP liveness and readiness probe.

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go
Comment (timebertt) [reply cmd/gardener-admission-controller/app/gardener_admission_controller.go:194]: Sorry, I have just now seen that you added the `HasSynced` funcs, that's great :)

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go]: I think `Kind` should always be uppercase, right?
Anyways, can you use `GetInformer` with an API object to get type safety?
```suggestion
	informer, err := k8sGardenClient.Cache().GetInformer(ctx, &corev1.Secret{})
```

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go]: `Shoot` is not in `corev1` 😄 
```suggestion
	informer, err = k8sGardenClient.Cache().GetInformer(ctx, &gardencorev1beta1.Shoot{})
```

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go]: I think the `HasSynced` funcs themselves don't wait, that's what `cache.WaitForCacheSync` is for, right?
```suggestion
	if !cache.WaitForCacheSync(ctx.Done(), hasSyncFuncs...) {
		return nil, errors.New("timed out waiting for caches to sync")
	}
```

**Review by timebertt — DISMISSED** [review]
Thanks for addressing my feedback.
A few more comments inline :)

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go
Comment (timuthy) [reply pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go]: Yes, you are right. I pushed it too early with the outstanding comments yesterday. I think we should also combine it with a readiness probe.

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml
Comment (timuthy) [reply charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml:83]: See discussion about health checks [here](https://github.com/gardener/gardener-extension-provider-gcp/pull/38#discussion_r387523877) and why we did not add them yet to any webhook component. Readiness is fine since we now wait for the cache sync.

**Comment by timuthy** [comment]
@rfranzke @timebertt let me address the last comments before you play around with it. I'll let you know asap.

**Comment by timuthy** [comment]
Thanks for you feedback which is now incorporated in the PR:
- Add `readyz` and `healthz` endpoints as well as readiness and liveness checks.
- Fixed wait for cache sync and informer retrieval bug.

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml
Comment (rfranzke) [inline charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml:73]: Any reason for 5m initial delay?

File: pkg/server/handler.go
Comment (rfranzke) [inline pkg/server/handler.go]: Can't you use https://github.com/gardener/gardener/blob/master/pkg/healthz/default.go like in https://github.com/gardener/gardener/blob/master/cmd/gardener-controller-manager/app/gardener_controller_manager.go#L346?

**Comment by timebertt** [comment]
No additional comments from my side.
Looks great :)

File: charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml
Comment (timuthy) [reply charts/gardener/controlplane/charts/runtime/templates/admission-controller/deployment.yaml:73]: Yes, it's the timeout for the cache sync wait. Otherwise we might already contact the `healthz` endpoint even though the initialization hasn't finished yet.

File: pkg/server/handler.go
Comment (timuthy) [reply pkg/server/handler.go]: Yes, I can use it as an alternative. @timebertt and I agreed that for now a "static" handler is sufficient but we might want to improve it in the future. So I decided to add a simple static handler for that in order to make it obvious to developer that no health checks at all are executed. WDYT?

Comment (rfranzke) [reply pkg/server/handler.go]: Maybe I'm missing something, but I don't see a difference in the static handler compared to this default healthz checker, or?

Comment (timuthy) [reply pkg/server/handler.go]: I'm thinking that we don't need no health checker with all its functionalities (start, set, stop, ...) if there are no real health checks happening. But I'm also happy to change it if it's beneficial.

Comment (timuthy) [reply pkg/server/handler.go]: I've changed it in dfc526810fe87b11ea8a0b9f0593841558b71091. Your call :)

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
Thanks @timuthy! Kindly update the pipeline definitions now that it's merged :)

**Pull request merged**
